ccmlib/cluster.py

```diff
@@ -121,6 +123,91 @@ def set_install_dir(self, install_dir=None, version=None, verbose=False):
 
         return self
 
+    def actively_watch_logs_for_error(self, on_error_call, interval=1):
+        """
+        Begins a thread that repeatedly scans system.log for new errors, every interval seconds.
+        (The first pass covers the entire log contents written at that point,
+        subsequent scans cover newly appended log messages).
+        Reports new errors, by calling the provided callback with an OrderedDictionary
+        mapping node name to a list of error lines.
+        Returns the thread itself, which should be .join()'ed to wrap up execution,
+        otherwise will run until the main thread exits.
+        """
+        class LogWatchingThread():
+            """
+            This class is embedded here for now, because it is used only from
+            within Cluster, and depends on cluster.nodelist().
+            """
+
+            def __init__(self, cluster):
+                self.executor = ThreadPoolExecutor(max_workers=1)
+                self.thread = None
+                self.cluster = cluster
+                self.req_stop_event = threading.Event()
+                self.done_event = threading.Event()
+                self.log_positions = defaultdict(int)
+
+            def start(self):
+                self.thread = self.executor.submit(self.run)
+
+            def scan(self):
+                errordata = OrderedDict()
+
+                try:
+                    for node in self.cluster.nodelist():
+                        scan_from_mark = self.log_positions[node.name]
+                        next_time_scan_from_mark = node.mark_log()
+                        if next_time_scan_from_mark == scan_from_mark:
+                            # log hasn't advanced, nothing to do for this node
+                            continue
+                        else:
+                            errors = node.grep_log_for_errors(from_mark=scan_from_mark)
+                        self.log_positions[node.name] = next_time_scan_from_mark
+                        if errors:
+                            errordata[node.name] = errors
+                except IOError as e:
+                    if 'No such file or directory' in str(e.strerror):
+                        pass  # most likely log file isn't yet written
+
+                    # in the case of unexpected error, report this thread to the callback
+                    else:
+                        errordata['log_scanner'] = [[str(e)]]
```

Comment:
Why need two level of nesting?

Reply:
I have now idea, It's a copy-paste from upstream

Reply:
seem like it's goes hand with their implementation of `grep_log_for_errors`
https://github.com/riptano/ccm/blob/master/ccmlib/node.py#L2199
and also to ours...

the reason isn't clear to me, but not going to change that now.